astro/src/content/blog/hundreds-millions-entities.mdx

```diff
@@ -0,0 +1,261 @@
+
+The entities were loaded one at a time. There is no way to bulk load entities in FusionAuth at this time. We saw loading rates of between 47,000 and 75,000 entities loaded per minute. They were loaded from a server in a different account but the same region as the HA server.
+
+The users were imported using the [User Import API](/docs/apis/users#import-users) in batches of 1000. The goal was not to stress test the number of users a FusionAuth deployment can support, but rather to have enough users to simulate decent scale.
+
+The [JWT Populate lambda](/docs/extend/code/lambdas/jwt-populate), which augmented the access token with entity data, looked like this:
+
+```javascript
+function populate(jwt, user, registration) {
+
+  // limited API key
+  var APIKEY='...';
+  
+  var response = fetch("http://localhost:9012/api/entity/grant/search?userId="+user.id, {
+    method: "GET",
+    headers: {
+      "Authorization": APIKEY
+    }
+  });
+
+  if (response.status === 200) {
+    var jsonResponse = JSON.parse(response.body);
+    var value = jsonResponse.grants[0].entity.data.attr2;
+    jwt.entityValue = value;
+  } else {
+    jwt.entityValue = "n/a";
+  }
+}
+```
+
+This code:
+
+* loads the first entity grant
+* extracts the second attribute of the associated entity, whatever it is
+* adds that attribute the token
+
+In a real world scenario you would have additional logic around the entities or attributes. You might add more of the values to the token. 
+
+## Setup Timeline
+
+The above setup took a fair bit of calendar time. We used an `m7g.2xlarge` instance running in the same region as the FusionAuth deployment, but in a different account. Due to experiments and other issues, the data loading took about 5 days to load ~98% of the entities. The users were created over 11 days, some of which overlapped with the entity loading. It took about 6 days to assign the users to the entities as outlined above. 
+
+Two bumps occurred during the load process. Both remind us that there is no such thing as "the cloud", only other people's computers.
+
+<img src="/img/blogs/hundreds-millions-entities/always-has-been.png" alt="Meme of astronauts saying the cloud has always been other people's computers" width="1200" />
+
+Yup, in both cases we ran out of disk space.
+
+<Aside type="note">
+The failures below occurred because this test system alerts were explicitly ignored by the FusionAuth support team.
+
+If this had been a customer deployment, the team would have been alerted well ahead of failure and taken steps to resolve the issues.
+</Aside>
+
+### Database Space
+
+First, we ran out of space in the relational database that is a key component of any FusionAuth instance. This happened at the point approximately 100,000,000 entities had been loaded. When this would happen in any other deployment depends on the size of the entities as well as other data loaded into FusionAuth.
```

Comment:
Would it be more accurate to say "When or if this would happen..."? As written it sounds like it's a certainty in any deployment.

Reply:
great catch.